engines/ic3sa.cpp

```diff
@@ -0,0 +1,880 @@
+    }
+  }
+
+  // HACK trouble to accurately identify predicates in boolector
+  // because of simplification. Something that starts as a predicate
+  // might not end up one, e.g. (= ((_ extract 0 0) x) #b0)
+  // just becomes ((_ extract 0 0) x) which we don't consider a predicate
+  // TODO possible fix is to just treat all 1-bit vals as predicates in btor
+  // TODO check if logging
+  if (solver_->get_solver_enum() == BTOR) {
+    return true;
+  }
+
+  Op op;
+  for (auto c : u.children) {
+    op = c->get_op();
+
+    // include bvnot for boolector (if we ever include boolector in this check)
+    if (op == Not || op == BVNot) {
+      c = smart_not(c);
+    }
+
+    // an equality is a predicate, so we just need to check
+    // is_predicate, not specifically for equalities
+    if (!is_predicate(c, boolsort, true)) {
+      return false;
+    }
+  }
+
+  // for now not checking the actual term (e.g. u.term)
+  // it's somewhat hard if the underlying solver does rewriting
+
+  // got through all checks without failing
+  return true;
+}
+
+void IC3SA::predecessor_generalization(size_t i,
+                                       const Term & c,
+                                       IC3Formula & pred)
+{
+  UnorderedTermSet all_coi_symbols = projection_set_;
+  justify_coi(ts_.next(c), all_coi_symbols);
+  assert(all_coi_symbols.size());
+
+  // get rid of next-state variables
+  UnorderedTermSet coi_symbols;
+  for (const auto & tt : all_coi_symbols) {
+    if (!ts_.is_next_var(tt)) {
+      coi_symbols.insert(tt);
+    }
+  }
+  assert(coi_symbols.size() <= ts_.statevars().size());
+  assert(coi_symbols.size());
+
+  logger.log(
+      2,
+      "IC3SA::generalize_predecessor projecting on {}/{} state variables",
+      coi_symbols.size(),
+      ts_.statevars().size());
+
+
+  UnorderedTermSet cube_lits;
+  // first populate with predicates
+  for (const auto & p : predset_) {
+    if (!in_projection(p, coi_symbols)) {
+      continue;
+    }
+
+    if (solver_->get_value(p) == solver_true_) {
+      cube_lits.insert(p);
+    } else {
+      cube_lits.insert(solver_->make_term(Not, p));
+    }
+  }
+
+  EquivalenceClasses ec = get_equivalence_classes_from_model(coi_symbols);
+  construct_partition(ec, cube_lits);
+  pred = ic3formula_conjunction(TermVec(cube_lits.begin(), cube_lits.end()));
+  assert(ic3formula_check_valid(pred));
+}
+
+void IC3SA::check_ts() const
+{
+  if (ts_.inputvars().size()) {
+    throw PonoException(
+        "IC3SA requires all state variables. Try option --promote-inputvars");
+  }
+
+  // TODO: add support for arrays
+
+  if (!conc_ts_.is_functional()) {
+    throw PonoException("IC3SA requires a functional transition system.");
+  }
+
+  for (const auto & sv : ts_.statevars()) {
+    SortKind sk = sv->get_sort()->get_sort_kind();
+    if (sk != BOOL && sk != BV)
+    {
+      throw PonoException("IC3SA currently only supports bit-vectors");
+    }
+  }
+  for (const auto & iv : ts_.inputvars()) {
```

Comment:
inputs are already checked here https://github.com/upscale-project/pono/pull/141/files#diff-d8b374ffc4535468eca246f9584183ee392cc728e6fd3a86332eeec176369206R176

why are we checking them again now?

Reply:
Good catch, fixed with b659b910a5913771d31f3fe6b0eb2b6b5041e109